File: backend/app/presentation_style.py
"""Shared presentation style defaults, theme palettes, and pagination helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .schemas import PresentationStylePayload


@dataclass(frozen=True)
class ThemePalette:
    """Resolved color palette for one presentation theme."""

    background: tuple[int, int, int]
    surface: tuple[int, int, int]
    border: tuple[int, int, int]
    header: tuple[int, int, int]
    accent: tuple[int, int, int]
    title_on_header: tuple[int, int, int]
    title_on_cover: tuple[int, int, int]
    body: tuple[int, int, int]
    subtitle: tuple[int, int, int]
    cover_background: tuple[int, int, int]


DEFAULT_PRESENTATION_STYLE = PresentationStylePayload()

THEME_LABELS: dict[str, str] = {
    "scholastic_blue": "学院蓝",
    "forest_green": "讲堂绿",
    "sunrise_orange": "晨光橙",
}

THEME_PALETTES: dict[str, ThemePalette] = {
    "scholastic_blue": ThemePalette(
        background=(247, 243, 234),
        surface=(255, 255, 255),
        border=(221, 207, 186),
        header=(24, 49, 73),
        accent=(199, 163, 100),
        title_on_header=(255, 255, 255),
        title_on_cover=(24, 49, 73),
        body=(35, 45, 58),
        subtitle=(90, 103, 118),
        cover_background=(246, 241, 231),
    ),
    "forest_green": ThemePalette(
        background=(243, 248, 241),
        surface=(255, 255, 255),
        border=(201, 221, 204),
        header=(30, 80, 67),
        accent=(180, 147, 91),
        title_on_header=(255, 255, 255),
        title_on_cover=(30, 80, 67),
        body=(34, 54, 48),
        subtitle=(88, 109, 100),
        cover_background=(237, 245, 237),
    ),
    "sunrise_orange": ThemePalette(
        background=(250, 244, 235),
        surface=(255, 255, 255),
        border=(232, 209, 185),
        header=(125, 70, 32),
        accent=(220, 140, 71),
        title_on_header=(255, 255, 255),
        title_on_cover=(125, 70, 32),
        body=(68, 49, 38),
        subtitle=(124, 97, 79),
        cover_background=(248, 239, 228),
    ),
}


def normalize_presentation_style(payload: Any) -> PresentationStylePayload:
    """Validate arbitrary metadata/request payload into a normalized style object."""
    if isinstance(payload, PresentationStylePayload):
        return payload
    if isinstance(payload, dict):
        return PresentationStylePayload.model_validate(payload)
    return DEFAULT_PRESENTATION_STYLE.model_copy()


def extract_presentation_style(metadata: Any) -> PresentationStylePayload:
    """Read the nested `presentation_style` block from plan metadata."""
    if isinstance(metadata, dict):
        return normalize_presentation_style(metadata.get("presentation_style"))
    return DEFAULT_PRESENTATION_STYLE.model_copy()


def get_theme_palette(theme_name: str | None) -> ThemePalette:
    """Resolve a palette, falling back to the default theme."""
    normalized = (theme_name or "").strip()
    return THEME_PALETTES.get(normalized, THEME_PALETTES[DEFAULT_PRESENTATION_STYLE.theme])


def resolve_density_limits(*, density: str, chars_per_line: int, max_lines: int, has_image_panel: bool) -> tuple[int, int]:
    """Adjust pagination limits to match the selected reading density."""
    normalized = (density or "").strip()
    if normalized == "comfortable":
        char_penalty = 4 if has_image_panel else 3
        line_penalty = 3 if has_image_panel else 4
        return max(chars_per_line - char_penalty, 12), max(max_lines - line_penalty, 4)
    if normalized == "compact":
        char_bonus = 2 if has_image_panel else 1
        line_bonus = 1 if has_image_panel else 2
        return chars_per_line + char_bonus, max_lines + line_bonus
    return chars_per_line, max_lines


def resolve_font_sizes(*, density: str, has_image_panel: bool) -> dict[str, int]:
    """Return a consistent font scale for the selected density."""
    normalized = (density or "").strip()
    if normalized == "compact":
        return {
            "title": 24,
            "cover_title": 28,
            "subtitle": 13,
            "body": 12 if has_image_panel else 13,
            "placeholder": 10,
            "branding": 10,
        }
    if normalized == "balanced":
        return {
            "title": 26,
            "cover_title": 30,
            "subtitle": 15,
            "body": 14 if has_image_panel else 15,
            "placeholder": 11,
            "branding": 11,
        }
    return {
        "title": 28,
        "cover_title": 32,
        "subtitle": 16,
        "body": 15 if has_image_panel else 17,
        "placeholder": 12,
        "branding": 12,
    }


def density_generation_hint(density: str) -> str:
    """Human prompt guidance for how aggressively the generator should split slides."""
    normalized = (density or "").strip()
    if normalized == "compact":
        return "内容密度偏好：compact。允许适度浓缩信息，但仍要保证投影可读，不要把长段落塞满整页。"
    if normalized == "balanced":
        return "内容密度偏好：balanced。优先保持课堂叙事顺畅，信息量和页数适中。"
    return "内容密度偏好：comfortable。宁可多拆几页，也不要缩小字号或让单页信息过满；图片页正文尽量控制在 2-4 行。"

File: frontend/lib/presentation-style.ts
import type { PresentationDensity, PresentationStylePayload, PresentationTheme } from "@/lib/types";

type ThemePalette = {
  background: string;
  surface: string;
  border: string;
  header: string;
  accent: string;
  titleOnHeader: string;
  titleOnCover: string;
  body: string;
  subtitle: string;
  coverBackground: string;
};

type FontScale = {
  title: number;
  coverTitle: number;
  subtitle: number;
  body: number;
  placeholder: number;
  branding: number;
};

type PreviewPage = Record<string, unknown> & {
  template: string;
  layout: string;
  title: string;
  body: string;
  bullet_points: string[];
  preview_page_number: number;
  preview_page_total: number;
  preview_original_index: number;
};

export const DEFAULT_PRESENTATION_STYLE: PresentationStylePayload = {
  theme: "scholastic_blue",
  density: "comfortable",
  school_name: null,
  logo_url: null,
  logo_file_id: null
};

export const THEME_LABELS: Record<PresentationTheme, string> = {
  scholastic_blue: "学院蓝",
  forest_green: "讲堂绿",
  sunrise_orange: "晨光橙"
};

export const DENSITY_LABELS: Record<PresentationDensity, string> = {
  comfortable: "舒展大字",
  balanced: "均衡",
  compact: "紧凑"
};

const THEME_PALETTES: Record<PresentationTheme, ThemePalette> = {
  scholastic_blue: {
    background: "#f7f3ea",
    surface: "#ffffff",
    border: "#ddcfba",
    header: "#183149",
    accent: "#c7a364",
    titleOnHeader: "#ffffff",
    titleOnCover: "#183149",
    body: "#232d3a",
    subtitle: "#5a6776",
    coverBackground: "#f6f1e7"
  },
  forest_green: {
    background: "#f3f8f1",
    surface: "#ffffff",
    border: "#c9ddcc",
    header: "#1e5043",
    accent: "#b4935b",
    titleOnHeader: "#ffffff",
    titleOnCover: "#1e5043",
    body: "#223630",
    subtitle: "#586d64",
    coverBackground: "#edf5ed"
  },
  sunrise_orange: {
    background: "#faf4eb",
    surface: "#ffffff",
    border: "#e8d1b9",
    header: "#7d4620",
    accent: "#dc8c47",
    titleOnHeader: "#ffffff",
    titleOnCover: "#7d4620",
    body: "#443126",
    subtitle: "#7c614f",
    coverBackground: "#f8efe4"
  }
};

const TEMPLATE_LIMITS: Record<string, { charsPerLine: number; maxLines: number }> = {
  title_body: { charsPerLine: 26, maxLines: 15 },
  title_body_image: { charsPerLine: 19, maxLines: 12 },
  title_subtitle: { charsPerLine: 24, maxLines: 4 }
};

function normalizeText(value: unknown) {
  return typeof value === "string" ? value.trim() : "";
}

function normalizeTheme(value: unknown): PresentationTheme {
  return value === "forest_green" || value === "sunrise_orange" || value === "scholastic_blue"
    ? value
    : DEFAULT_PRESENTATION_STYLE.theme;
}

function normalizeDensity(value: unknown): PresentationDensity {
  return value === "balanced" || value === "compact" || value === "comfortable"
    ? value
    : DEFAULT_PRESENTATION_STYLE.density;
}

function normalizeTemplate(value: unknown, layout: unknown) {
  const template = normalizeText(value).toLowerCase();
  const normalizedLayout = normalizeText(layout).toLowerCase();
  if (template === "title_subtitle" || normalizedLayout === "cover" || normalizedLayout === "closing") {
    return "title_subtitle";
  }
  if (template === "title_body_image" || normalizedLayout === "image") {
    return "title_body_image";
  }
  return "title_body";
}

function bodyToBullets(body: string) {
  return body
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean);
}

function wrapParagraph(paragraph: string, width: number) {
  const lines: string[] = [];
  let current = "";
  for (const char of paragraph) {
    current += char;
    if (current.length >= width) {
      lines.push(current);
      current = "";
    }
  }
  if (current) {
    lines.push(current);
  }
  return lines.length ? lines : [paragraph];
}

function splitLinesIntoPages(lines: string[], maxLines: number) {
  const pages: string[][] = [];
  let current: string[] = [];

  lines.forEach((line) => {
    if (current.length && current.length + 1 > Math.max(maxLines, 1)) {
      pages.push(current);
      current = [];
    }
    current.push(line);
  });

  if (current.length) {
    pages.push(current);
  }

  return pages;
}

function rebalanceSparseTailPages(pageLines: string[][], maxLines: number) {
  if (pageLines.length < 2) {
    return pageLines;
  }

  const visibleTailLines = pageLines[pageLines.length - 1].filter((line) => line.trim()).length;
  const sparseThreshold = Math.max(2, Math.min(3, Math.floor(maxLines / 3)));
  if (visibleTailLines > sparseThreshold) {
    return pageLines;
  }

  const flatLines = pageLines.flat();
  if (flatLines.length <= maxLines) {
    return pageLines;
  }

  const rebalanced: string[][] = [];
  let cursor = 0;
  let remainingLines = flatLines.length;

  for (let pageIndex = 0; pageIndex < pageLines.length; pageIndex += 1) {
    const pagesLeft = pageLines.length - pageIndex;
    const pageSize = Math.min(maxLines, Math.max(Math.ceil(remainingLines / pagesLeft), 1));
    rebalanced.push(flatLines.slice(cursor, cursor + pageSize));
    cursor += pageSize;
    remainingLines -= pageSize;
  }

  return rebalanced;
}

function paginateSlideText(text: string, charsPerLine: number, maxLines: number) {
  const normalizedCharsPerLine = Math.max(charsPerLine, 1);
  const normalizedMaxLines = Math.max(maxLines, 1);
  const paragraphs = String(text || "")
    .split("\n")
    .map((item) => item.trim());
  const lines: string[] = [];

  paragraphs.forEach((paragraph) => {
    if (!paragraph) {
      if (lines.length && lines[lines.length - 1] !== "") {
        lines.push("");
      }
      return;
    }
    lines.push(...wrapParagraph(paragraph, normalizedCharsPerLine));
  });

  if (!lines.length) {
    return [""];
  }

  const splitPages = splitLinesIntoPages(lines, normalizedMaxLines);
  const singlePagePreferred = preferSinglePageForLightOverflow(splitPages, normalizedCharsPerLine, normalizedMaxLines);
  const pageLines = rebalanceSparseTailPages(singlePagePreferred, normalizedMaxLines);
  const pages = pageLines.map((page) => page.join("\n").trim()).filter(Boolean);

  return pages.filter(Boolean).length ? pages.filter(Boolean) : [String(text || "").trim()];
}

function preferSinglePageForLightOverflow(pageLines: string[][], charsPerLine: number, maxLines: number) {
  if (pageLines.length !== 2) {
    return pageLines;
  }

  const [firstPage, secondPage] = pageLines;
  const visibleFirstLines = firstPage.filter((line) => line.trim()).length;
  const visibleSecondLines = secondPage.filter((line) => line.trim()).length;
  const totalVisibleLines = visibleFirstLines + visibleSecondLines;
  if (visibleSecondLines === 0) {
    return [firstPage];
  }
  if (visibleSecondLines > Math.max(2, Math.min(3, Math.floor(maxLines / 4)))) {
    return pageLines;
  }
  if (totalVisibleLines > maxLines + 1) {
    return pageLines;
  }
  if (totalVisibleLines > Math.min(maxLines + 1, 6)) {
    return pageLines;
  }

  const firstPageText = firstPage.map((line) => line.trim()).join("");
  const secondPageText = secondPage.map((line) => line.trim()).join("");
  if (!secondPageText) {
    return [firstPage];
  }

  const overflowRatio = secondPageText.length / Math.max(charsPerLine, 1);
  if (visibleFirstLines >= maxLines - 1 && overflowRatio > 0.6) {
    return pageLines;
  }

  return [[...firstPage, ...secondPage]];
}

function resolveDensityLimits(density: PresentationDensity, charsPerLine: number, maxLines: number, hasImagePanel: boolean) {
  if (density === "comfortable") {
    const charPenalty = hasImagePanel ? 4 : 3;
    const linePenalty = hasImagePanel ? 3 : 4;
    return {
      charsPerLine: Math.max(charsPerLine - charPenalty, 12),
      maxLines: Math.max(maxLines - linePenalty, 4)
    };
  }
  if (density === "compact") {
    const charBonus = hasImagePanel ? 2 : 1;
    const lineBonus = hasImagePanel ? 1 : 2;
    return {
      charsPerLine: charsPerLine + charBonus,
      maxLines: maxLines + lineBonus
    };
  }
  return { charsPerLine, maxLines };
}

export function normalizePresentationStyle(style: unknown): PresentationStylePayload {
  if (!style || typeof style !== "object" || Array.isArray(style)) {
    return { ...DEFAULT_PRESENTATION_STYLE };
  }
  const payload = style as Record<string, unknown>;
  return {
    theme: normalizeTheme(payload.theme),
    density: normalizeDensity(payload.density),
    school_name: normalizeText(payload.school_name) || null,
    logo_url: normalizeText(payload.logo_url) || null,
    logo_file_id: normalizeText(payload.logo_file_id) || null
  };
}

export function extractPresentationStyle(metadata: unknown) {
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return normalizePresentationStyle(null);
  }
  return normalizePresentationStyle((metadata as Record<string, unknown>).presentation_style);
}

export function getThemePalette(theme: PresentationTheme) {
  return THEME_PALETTES[theme] || THEME_PALETTES[DEFAULT_PRESENTATION_STYLE.theme];
}

export function resolveFontSizes(density: PresentationDensity, hasImagePanel: boolean): FontScale {
  if (density === "compact") {
    return {
      title: 24,
      coverTitle: 28,
      subtitle: 13,
      body: hasImagePanel ? 12 : 13,
      placeholder: 10,
      branding: 10
    };
  }
  if (density === "balanced") {
    return {
      title: 26,
      coverTitle: 30,
      subtitle: 15,
      body: hasImagePanel ? 14 : 15,
      placeholder: 11,
      branding: 11
    };
  }
  return {
    title: 28,
    coverTitle: 32,
    subtitle: 16,
    body: hasImagePanel ? 15 : 17,
    placeholder: 12,
    branding: 12
  };
}

export function slideUsesImagePanel(slide: Record<string, unknown>) {
  return normalizeTemplate(slide.template, slide.layout) === "title_body_image";
}

export function paginateSlidesForPreview(
  slides: Array<Record<string, unknown>>,
  style: PresentationStylePayload
): PreviewPage[] {
  const pages: PreviewPage[] = [];

  slides.forEach((slide, slideIndex) => {
    const template = normalizeTemplate(slide.template, slide.layout);
    const title = normalizeText(slide.title) || `第 ${slideIndex + 1} 页`;
    const subtitle = normalizeText(slide.subtitle);
    const body =
      normalizeText(slide.body) ||
      (Array.isArray(slide.bullet_points) ? slide.bullet_points.filter((item) => typeof item === "string").join("\n") : "");

    if (template === "title_subtitle") {
      pages.push({
        ...slide,
        template,
        layout: "cover",
        title,
        subtitle: subtitle || body,
        body: "",
        bullet_points: [],
        preview_page_number: 1,
        preview_page_total: 1,
        preview_original_index: slideIndex
      });
      return;
    }

    const base = TEMPLATE_LIMITS[template] || TEMPLATE_LIMITS.title_body;
    const hasImagePanel = template === "title_body_image";
    const limits = resolveDensityLimits(style.density, base.charsPerLine, base.maxLines, hasImagePanel);
    const chunks = paginateSlideText(body, limits.charsPerLine, limits.maxLines);

    chunks.forEach((chunk, pageIndex) => {
      pages.push({
        ...slide,
        template,
        layout: hasImagePanel ? "image" : "title_content",
        title: chunks.length > 1 ? `${title}（${pageIndex + 1}/${chunks.length}）` : title,
        body: chunk,
        bullet_points: bodyToBullets(chunk),
        preview_page_number: pageIndex + 1,
        preview_page_total: chunks.length,
        preview_original_index: slideIndex
      });
    });
  });

  return pages;
}
